### PR TITLE
spack module : reworking of the command

### PR DESCRIPTION
**Pull request by alalazo**
##### Rationale
I started this because at EPFL we need a way to regenerate module files in production with a bit more control over what gets changed than is currently available.

##### New sub-commands 
- [x] reworked the extensions to `spack module find` in #721 into `spack module loads`
- [x] added `spack module rm` to delete module files

##### `spack module refresh`
- [x] `spack module refresh` asks for confirmation before refreshing
- [x] deleting the module file tree is now optional

##### modules.yaml
- [x] added a new keyword to `modules.yaml` that influences module file name : hash_length

##### Fixes
- [x] reintroduced `path` in `EnvModules` : was lost in a15aee5d8e7bcdd37d9ee25fcf6ded449ef305d5 and prevents `spack module refresh` from working

##### Other modifications
- [x] added common cli options in `spack.cmd.commo.arguments` to be reused by multiple commands
- [x] any `spack module` subcommand accepts a constraint to select subsets of packages
- [x] `spack module` delete operations now asks user confirmation by default (can be turned off with `-y`)
- [x] `EnvModules` instances can now skip writing on file name clashes
- [x] add unit tests to improve the coverage for this command

@tgamblin @glennpj @robertdfrench just pinging you as I think you might be interested in this module related changeset

### CODE REVIEW

**Comment by mwilliammyers** [comment]
@alalazo Are you planning on adding an `-f` option to `refresh`?

**Comment by alalazo** [comment]
@mwilliammyers Do you mean `-f` like in `spack edit -f` to force overwriting of files? In that case I think it is implied as this command rewrites module files most part f the time... or you had something else in mind?

**Comment by mwilliammyers** [comment]
@alalazo I guess I thought you had written that if it was going to overwrite the module then it would prompt for confirmation (in which case a `-f` for force would be handy) but I think I was thinking about the other issue ticket. My apologies. 

I really like this PR, I have on many occasions wanted to only refresh certain modules.

File: bin/spack [hunk 1]
```diff
@@ -77,7 +77,7 @@ import llnl.util.tty as tty
 from llnl.util.tty.color import *
 import spack
 from spack.error import SpackError
-import argparse
+from external import argparse
```
Comment (adamjstewart) [inline bin/spack]: See #1030

Comment (alalazo) [inline bin/spack]: @adamjstewart Ouch, this went in accidentally. Thanks for pointing this out.

**Comment by alalazo** [comment]
@mathstuf Not a problem to change the explicit recursion to a call to `traverse`.

@citibeth Can you confirm that this works as expected in #721 ? 

BTW, I changed the default of `args.prefix` to be the empty string (for me it was literally printing 'None' if called with only `-s`). I was also thinking of adding logic to skip blacklisted modules, otherwise you may end up with faulty lines in the script (depending on what is in `modules.yaml`).

**Comment by citibeth** [comment]
@alalazo I think this is a good opportunity to fix the things that annoyed me about #721.  I will gladly adjust to any compatibility-breaking improvements made at this point.

First I will describe how I use it.  I have the following script, which generates `~/env/icebin`:
```
SPACKENV=$HOME/env/icebin
FIND='spack module find --shell tcl'

echo '#!/bin/sh -f' >$SPACKENV
echo '# ---- Machine generated; do not edit!' >>$SPACKENV
echo 'module purge' >>$SPACKENV

$FIND git emacs ncview nco py-rtree >$SPACKENV
$FIND everytrace@develop >>$SPACKENV

$FIND --dependencies icebin@0.1.4 >>$SPACKENV
$FIND --dependencies py-giss py-proj >>$SPACKENV```
```
Then I source `~/env/icebin` when I want to do IceBin-related stuff.  (I have other `env` scripts set up as well).  Things to note here:

1. I use `--dependencies` on anything Python-related, because `RPATH` doesn't work for Python code --- which requires `PYTHONPATH`, the evil Python twin of `LD_LIBRARY_PATH`.  Compiled Python modules also don't get built with `RPATH`, even though they are binaries that COULD use `RPATH`.  Both these sins are covered by `--dependencies`.

2. The `--shell tcl` is obnoxious.  The overall command is now too long: `spack module find --shell tcl`.  Not only that, but I always think `tcl` is one of the packages I'm trying to find.  Maybe we need a different sub-verb that means `find --shell tcl`.  For example: `spack module env`, because it is used to create environments.

3. The command can take multiple specs (as demonstrated in the above script), at least if you're not using '--dependencies'.  I added this feature to try to speed things up.  It wasn't very effective at that, and @tgamblin has introduced some better speed improvements.  It has also complicated life in subtle ways.  For example, Spack will quit on the first thing it can't find, and you won't know what else it couldn't find.  Finally, I don't know if the multi-spec stuff made it into the final merge.  If it's there, feel free to remove it.  If not... I wouldn't bother adding it back.

**Comment by alalazo** [comment]
@citibeth I followed your suggestion and added a new subcommand `load-list` and made `tcl` the default (so there's no need to specify it) : feel free to suggest a better name or a better description than the one that is currently in place if it doesn't convey the purpose of the thing.

Now you can query on a single set of constraints, e.g. : 
```
$ spack module load-list   ^lua
lmod-6.4.1-gcc-4.8
lua-luafilesystem-1_6_3-gcc-4.8
lua-luaposix-33.4.0-gcc-4.8

$ spack module load-list -s   ^lua
# lmod@6.4.1%gcc@4.8 arch=linux-Ubuntu14-x86_64
module load lmod-6.4.1-gcc-4.8
# lua-luafilesystem@1_6_3%gcc@4.8 arch=linux-Ubuntu14-x86_64
module load lua-luafilesystem-1_6_3-gcc-4.8
# lua-luaposix@33.4.0%gcc@4.8 arch=linux-Ubuntu14-x86_64
module load lua-luaposix-33.4.0-gcc-4.8

$ spack module load-list -s -r  ^lua
# lua@5.3.2%gcc@4.8 arch=linux-Ubuntu14-x86_64
module load lua-5.3.2-gcc-4.8
# lua-luafilesystem@1_6_3%gcc@4.8 arch=linux-Ubuntu14-x86_64
module load lua-luafilesystem-1_6_3-gcc-4.8
# lua-luaposix@33.4.0%gcc@4.8 arch=linux-Ubuntu14-x86_64
module load lua-luaposix-33.4.0-gcc-4.8
# lmod@6.4.1%gcc@4.8 arch=linux-Ubuntu14-x86_64
module load lmod-6.4.1-gcc-4.8

$ spack module load-list -s -r -p 'PREFIX ' ^lua
# lua@5.3.2%gcc@4.8 arch=linux-Ubuntu14-x86_64
module load PREFIX lua-5.3.2-gcc-4.8
# lua-luafilesystem@1_6_3%gcc@4.8 arch=linux-Ubuntu14-x86_64
module load PREFIX lua-luafilesystem-1_6_3-gcc-4.8
# lua-luaposix@33.4.0%gcc@4.8 arch=linux-Ubuntu14-x86_64
module load PREFIX lua-luaposix-33.4.0-gcc-4.8
# lmod@6.4.1%gcc@4.8 arch=linux-Ubuntu14-x86_64
module load PREFIX lmod-6.4.1-gcc-4.8
mculpo@nuvolari:~/PycharmProjects/spack$ 
```
Does this help with `PYTHONPATH`?

File: lib/spack/spack/cmd/module.py [hunk 2]
```diff
@@ -23,135 +23,206 @@
...
+        # FIXME : during module file creation nodes seem to be visited multiple
+        # FIXME : times even if cover='nodes' is given. This work around permits
+        # FIXME : to get a unique list of spec anyhow. Do we miss a merge
+        # FIXME : step among nodes that refer to the same package?
+        # FIXME : (same problem as in spack/modules.py)
+        seen = set()
+        seen_add = seen.add
+        for spec in specs_from_user_constraint:
+            specs.extend(
+                [item for item in spec.traverse(order='post', cover='nodes') if not (item in seen or seen_add(item))]
```
Comment (alalazo) [inline lib/spack/spack/cmd/module.py]: @mathstuf Better like this ?

Comment (mathstuf) [inline lib/spack/spack/cmd/module.py]: Yep, thanks :) .

Comment (alalazo) [inline lib/spack/spack/cmd/module.py]: @mathstuf @tgamblin BTW : do you know anything about the FIXME above ? I noticed that in many contexts if I call `traverse` with `cover=nodes` I still get duplicate nodes... this should not happen according to `__doc__` and is the cause of the `seen` hack...

**Comment by citibeth** [comment]
On Fri, Jul 1, 2016 at 6:48 AM, Massimiliano Culpo <notifications@github.com
> wrote:

> @citibeth <https://github.com/citibeth> I followed your suggestion and
> added a new subcommand load-list
>
Thanks!  Might I suggest calling it simply `loads`?

> and made tcl the default (so there's no need to specify it)
>
That's good news, since Dotkit is not really used outside LLNL.


> Now you can query on a single set of constraints, e.g. :
>
> $ spack module load-list   ^lua
> lmod-6.4.1-gcc-4.8
> lua-luafilesystem-1_6_3-gcc-4.8
> lua-luaposix-33.4.0-gcc-4.8
>
>
Hmm... I think that what's currently `-s` should probably be the default,
and the non-dash-s option should be made the flagged option.  The reason
is, why would you want just the list when we already have `spack load`?  It
would before something like:
```
module load `spack module loads ^lua`
```
(PS: I'm unfamiliar with `^lua` sytax.  Does this mean, everything that
depends on `lua`?)

>
> $ spack module load-list -s -r  ^lua
> # lua@5.3.2%gcc@4.8 arch=linux-Ubuntu14-x86_64
> module load lua-5.3.2-gcc-4.8
> # lua-luafilesystem@1_6_3%gcc@4.8 arch=linux-Ubuntu14-x86_64
> module load lua-luafilesystem-1_6_3-gcc-4.8
> # lua-luaposix@33.4.0%gcc@4.8 arch=linux-Ubuntu14-x86_64
> module load lua-luaposix-33.4.0-gcc-4.8
> # lmod@6.4.1%gcc@4.8 arch=linux-Ubuntu14-x86_64
> module load lmod-6.4.1-gcc-4.8
>
>
As per previous discussion, `--dependencies` is the correct flag here.  I
believe we did decide on `-r` as the short version of it, as shown above.
But we should make sure the decision wasn't `-R`.

>
> $ spack module load-list -s -r -p 'PREFIX ' ^lua
> # lua@5.3.2%gcc@4.8 arch=linux-Ubuntu14-x86_64
> module load PREFIX lua-5.3.2-gcc-4.8
> # lua-luafilesystem@1_6_3%gcc@4.8 arch=linux-Ubuntu14-x86_64
> module load PREFIX lua-luafilesystem-1_6_3-gcc-4.8
> # lua-luaposix@33.4.0%gcc@4.8 arch=linux-Ubuntu14-x86_64
> module load PREFIX lua-luaposix-33.4.0-gcc-4.8
> # lmod@6.4.1%gcc@4.8 arch=linux-Ubuntu14-x86_64
> module load PREFIX lmod-6.4.1-gcc-4.8
> mculpo@nuvolari:~/PycharmProjects/spack$
>
> Does this help with PYTHONPATH?
>
Nice idea and good try, but unfortunately not quite.  There will still be
`.so` libraries that some of the Python modules depend on, that are not
loaded.  (Eg: `py-netcdf4` needs `netcdf` to be loaded).

-- Elizabeth

**Comment by alalazo** [comment]
@tgamblin Can you have a look at the last commit here and tell me if you like the idea of aggregating commonly used cli options for reuse across different commands ? If it is fine with you I'll proceed cleaning this PR up and adding tests.

If you don't like the (relatively simple) design or you think it needs more work to get it right I can extract from this PR : 
- the addition of `hash_length` to `modules.yaml`
- the fix to the currently broken `spack module refresh` (see description above)

and submit them as separate, smaller PRs

**Comment by alalazo** [comment]
@tgamblin (or anybody else willing to try this out) : the PR is ready for review

**Comment by glennpj** [comment]
@alalazo Is there a way to _not_ generate the entries from prefix_inspections in the module file for specific packages? The reason I ask is that the modules for the intel compilers are wrong with the auto scanning. I would like to fix the module with explicit entries but the auto generated values are in the way.

Thanks.

**Comment by glennpj** [comment]
@alalazo I think what is needed is a way to turn off the inspections in the package.py file, either all or selected ones, so that the module file generation uses the values set in the package `setup_environment` function only. I understand this would be different PR if that needs to be added. I was hoping I was just missing how to do that.

Should I submit a PR (marked WIP) for the intel-parallel-studio package with the added `setup_environment` so there will be a reference/test case?

Thanks.

**Comment by glennpj** [comment]
@alalazo The `module refresh` command is working again and I see the module files created with a shorter hash. So far I have tried refreshing a single module and removing modules. The --delete-tree option should probably be made inactive when specifying a constraint. Also, I can not get multiple constraints to work.
```
spack module refresh -h
usage: spack module refresh [-h] [--delete-tree] [-m {tcl,dotkit}] [-y]
                            [constraint [constraint ...]]
```
That would seem to indicate that multiple constraints can be given but it does not seem to work.
```
spack module refresh zlib jpeg
==> No package matches your query
```

Thanks.

**Comment by alalazo** [comment]
@glennpj 

> The --delete-tree option should probably be made inactive when specifying a constraint.

I see your point here. Anyhow I don't know if it is worth to add some logic just to disable it in that case : the `--delete-tree` is always explicit and has not short version by choice, and that should make it difficult to just add it by mistake. 

> Also, I can not get multiple constraints to work.

**TLDR** : the command is working as I expect, but I should write better help on how to use / what is a constraint.

I guess that is my fault in not being able to convey how constraints are met. Basically doing : 
```
spack module refresh zlib jpeg
```
you are asking to refresh `tcl` module for packages that meet BOTH the constraints `zlib` and `jpeg` (and obviously the number of those is zero). The idea at the base of constraints is to be able to say something like : 
```
spack module refresh ^python %gcc
```
meaning : refresh packages that depends on python and are compiled with gcc. It should work the same way as it works in `modules.yaml`. 

On a sidenote, none of the commands that were accepting multiple specs (I use specs here instead of constraints to mark the difference in semantics between the two) seems to be working after #561, at least for me :
```
$ spack install zlib jpeg
==> Error: There are no valid versions for zlib that match ''
```
I honestly thought that that was a change in the command due to the fact that now specs requires using spaces, but as you raise this issue, I wonder if it is not then a bug in the parser... @tgamblin ?

**Comment by alalazo** [comment]
@citibeth 

> Nice idea and good try, but unfortunately not quite.  There will still be
`.so` libraries that some of the Python modules depend on, that are not
loaded.  (Eg: `py-netcdf4` needs `netcdf` to be loaded).

Well, right now with this PR you can do something like : 
```
$ spack module loads ^lua && spack module loads hdf5+mpi
# lmod@6.4.1%gcc@4.8 arch=linux-Ubuntu14-x86_64
module load lmod-6.4.1-gcc-4.8-6n2nymi
# lua-luafilesystem@1_6_3%gcc@4.8 arch=linux-Ubuntu14-x86_64
module load lua-luafilesystem-1_6_3-gcc-4.8-s7jcpfx
# lua-luaposix@33.4.0%gcc@4.8 arch=linux-Ubuntu14-x86_64
module load lua-luaposix-33.4.0-gcc-4.8-wekp5dh
# hdf5@1.8.16%gcc@4.8~cxx~debug~fortran+mpi+shared~szip~threadsafe arch=linux-Ubuntu14-x86_64
module load hdf5-1.8.16-gcc-4.8-7f5io6x
```

If this PR gets accepted a thing that I would like to add is the possibility to parse logical constraints expression from the command line, meaning that I would like : 
```
$ spack module loads ^lua or hdf5+mpi
# lmod@6.4.1%gcc@4.8 arch=linux-Ubuntu14-x86_64
module load lmod-6.4.1-gcc-4.8-6n2nymi
# lua-luafilesystem@1_6_3%gcc@4.8 arch=linux-Ubuntu14-x86_64
module load lua-luafilesystem-1_6_3-gcc-4.8-s7jcpfx
# lua-luaposix@33.4.0%gcc@4.8 arch=linux-Ubuntu14-x86_64
module load lua-luaposix-33.4.0-gcc-4.8-wekp5dh
# hdf5@1.8.16%gcc@4.8~cxx~debug~fortran+mpi+shared~szip~threadsafe arch=linux-Ubuntu14-x86_64
module load hdf5-1.8.16-gcc-4.8-7f5io6x
```
to give the same result as above.

**Comment by glennpj** [comment]
@alalazo 
I think what you refer to as constraints I thought were specs. The following in the help blurb would have clear to me.
```
spack module refresh -h
usage: spack module refresh [-h] [--delete-tree] [-m {tcl,dotkit}] [-y]
                            [module [spec]]
```

Regarding the `--delete-tree` option:
With the ability to add suffixes to module names in #984 and variable hash length in this PR, the `module refresh` command can create additional modules for a package in addition to refreshing the content. This means the module tree will have duplicate modules with different hash lengths and suffixes pointing to the same package. In the global case this can be cleaned up with `--delete-tree` but not for the single module case. That would require use of the `rm` subcommand and would be a two step process. 

One suggestion would be to replace the `--delete-tree` option with a `--remove-duplicates` (or similar) option. In the global refresh case that would do the same thing as `--delete-tree` but for the single module case have it remove any duplicate module names of that module [constraint/spec].

For example, after running `module refresh` with the new hash length.
```
module avail bzip2
bzip2/1.0.6-gcc-4.8.5-rgrn5su
bzip2/1.0.6-gcc-4.8.5-rgrn5su5tk2kijygfzndddpm47opvjzv
```
Those are exactly the same, just the hash length is different. It may be useful to keep the original around, in case people refer to it in a script, so the default should be not to remove it. But it may also be useful to remove those to avoid confusion and thus doing something like
```
spack module refresh --remove-duplicate bzip2
```
would be useful. That is a contrived example as I would probably want to do that globally but that would make the behavior consistent.

Thanks.

**Comment by glennpj** [comment]
@alalazo I see your point; it would be complicated.

I have been testing this PR a bit more and it seems quite good to me. I intentionally created a collision by removing the hashes and the operation correctly aborted with a message about the collision and listed the specs. I took that information, added a suffix option to clear the collision and regenerated the modules. Quite nice.

I also tested when a new package installation would generate a module file that collided with an existing module. In this case, a warning was printed with a reference to the existing module file and the full spec of the newly installed package.
```
==> Warning: Module file already exists : skipping creation
```
I wonder if that should be an error rather than just a warning. If that were a dependency getting built the fact that the environment module creation was skipped may be missed. I think showing the same error as the `module refresh` command does would be helpful there.

**Comment by alalazo** [comment]
@tgamblin Synced with `develop` and ready for review

**Comment by alalazo** [comment]
@tgamblin @becker33 Re-synced with `develop`

File: lib/spack/spack/cmd/module.py (hunk 2 above)
Comment (mathstuf) [inline lib/spack/spack/cmd/module.py]: It looks like by default, the comparison operator for specs is used. You can use `key='name'` to use the name of the spec as the uniqueness test.

Comment (mathstuf) [inline lib/spack/spack/cmd/module.py]: Actually, looks like there's a bug in it. Filing a PR now.

Comment (mathstuf) [inline lib/spack/spack/cmd/module.py]: Heh, when it's fixed, the test suite is unhappy :) . Seems the tests were written to the behavior, not the docs.

Comment (mathstuf) [inline lib/spack/spack/cmd/module.py]: Ah, nope. `traverse` is right. I think the comment probably only applies to concretized specs, otherwise the DAG does not combine its duplicate specs (by name).

File: lib/spack/spack/cmd/module.py [hunk 3]
```diff
@@ -23,135 +23,233 @@
...
+    loads_parser.add_argument(
+        '-p', '--prefix', dest='prefix', default='',
+        help='Prepend to module names when issuing module load commands'
+    )
+    arguments.add_common_arguments(
+        loads_parser, ['constraint', 'module_type', 'recurse_dependencies']
+    )
+
+
+class MultipleMatches(Exception):
```
Comment (tgamblin) [inline lib/spack/spack/cmd/module.py]: These always get caught in this so it's probably ok, but why not make them extend `SpackException` so they can be handled gracefully if they ever get out?

**Comment by tgamblin** [comment]
@alalazo: looks good!  Thanks!

File: lib/spack/spack/cmd/module.py (hunk 3 above)
Comment (alalazo) [inline lib/spack/spack/cmd/module.py]: @tgamblin will do in a future PR :+1:

**Comment by mwilliammyers** [comment]
@alalazo I noticed this PR added an option for `hash_length` in the case that this is 0, is there currently a mechanism to detect if a modulefile will be overwritten with a modulefile for one with a different DAG?

**Comment by alalazo** [comment]
@mwilliammyers Yes. If you try it should either : 
- report a warning at install time
- fail with an error on module refresh

see @glennpj [comment](https://github.com/LLNL/spack/pull/1135#issuecomment-230598474) on this.

@glennpj The warning behavior may be changed if people don't like it. I just made it like that because I thought that an install failure (with consequent removal of the prefix) for a package that takes a long time to build shouldn't happen for a name collision in a module file. 

I already pictured people cursing at me if that would have happened after a succesfull `trilinos` build :smile:
